src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs

```diff
@@ -307,128 +402,148 @@ public static (uint encoded, int bitLength) Encode(int data)
+                        }
+                        bitsInAcc -= 8;
+                    }
+                } while (bitsInAcc >= 8);
             }
 
-            return j;
-        }
+            // Finish decoding last < 8 bits of src.
+            // Although bellow code could be inserted in above do..while,
+            // processing of last byte has to handle several corner cases was decided, for performance reasons,
+            // to not have in main decoding loop.
+            while (bitsInAcc > 0)
+            {
+                Debug.Assert(bitsInAcc < 8);
 
-        /// <summary>
-        /// Decodes a single symbol from a 32-bit word.
-        /// </summary>
-        /// <param name="data">A 32-bit word containing a Huffman encoded symbol.</param>
-        /// <param name="validBits">
-        /// The number of bits in <paramref name="data"/> that may contain an encoded symbol.
-        /// This is not the exact number of bits that encode the symbol. Instead, it prevents
-        /// decoding the lower bits of <paramref name="data"/> if they don't contain any
-        /// encoded data.
-        /// </param>
-        /// <param name="decodedBits">The number of bits decoded from <paramref name="data"/>.</param>
-        /// <returns>The decoded symbol.</returns>
-        internal static int DecodeValue(uint data, int validBits, out int decodedBits)
-        {
-            // The code below implements the decoding logic for a canonical Huffman code.
-            //
-            // To decode a symbol, we scan the decoding table, which is sorted by ascending symbol bit length.
-            // For each bit length b, we determine the maximum b-bit encoded value, plus one (that is codeMax).
-            // This is done with the following logic:
-            //
-            // if we're at the first entry in the table,
-            //    codeMax = the # of symbols encoded in b bits
-            // else,
-            //    left-shift codeMax by the difference between b and the previous entry's bit length,
-            //    then increment codeMax by the # of symbols encoded in b bits
-            //
-            // Next, we look at the value v encoded in the highest b bits of data. If v is less than codeMax,
-            // those bits correspond to a Huffman encoded symbol. We find the corresponding decoded
-            // symbol in the list of values associated with bit length b in the decoding table by indexing it
-            // with codeMax - v.
+                // Check for correct EOS, which is padding with ones till end of byte
+                // when we STARTED new huffman code in last 8 bits (lookupTableIndex was reset to 0 -> root lookup table).
+                if (lookupTableIndex == 0)
+                {
+                    // Check if all remaining bits are ones.
+                    uint ones = uint.MaxValue >> (64 - bitsInAcc);
```

Comment:
Is this correct? Type is `uint` (32 bits) but you're rotating by 57-63 positions. I [works](https://dotnetfiddle.net/9kXObi), don't get me wrong, but it's just thanks to this C# feature: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/bitwise-and-shift-operators#shift-count-of-the-shift-operators.

Reply:
Good catch. It is lucky bug.
Fixed by replacing 64 => 32.
BTW, that C# feature https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/bitwise-and-shift-operators#shift-count-of-the-shift-operators is quite controversial to me as it differes from C/C++ behavior. the fact that `((ulong)0x1 >> 64) == 0x1` is not very intuitive.